### ISSUE TITLE
New Rule: Duplicate Aliases in Join clause or Select Clause

### ISSUE THREAD

**Issue by alanmcruickshank** [sqlfluff/sqlfluff#3731]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

Occasionally I see queries like:

```sql
SELECT
    a as foo,
    many,
    other,
    columns,
    b as foo
FROM table
```

The is usually user error in defining two columns with the same alias. There's an equivalent issue in join clause aliases. This wouldn't be a fixable violation, but something good to highlight to the user as the sign of a likely bug.

### Dialect

all

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
